src/encoded/types/file.py

```diff
@@ -60,3 +92,134 @@ def upload_credentials(self) -> Union[str, None]:
+             permission='edit', validators=[schema_validator({"type": "object"})])
+@debug_log
+def post_upload(context, request):
+    return CorePostUpload(context, request)
+
+
+@view_config(name='download', context=File, request_method='GET',
+             permission='view', subpath_segments=[0, 1])
+def download(context, request):
+    return CoreDownload(context, request)
+
+
+def validate_processed_file_unique_md5_with_bypass(context, request):
+    """ validator to check md5 on output files, unless you tell it not to
+        This validator is duplicated from encoded-core because of processed file rename
+    """
+    # skip validator if not file output
+    if context.type_info.item_type != 'output_file':
+        return
+    data = request.json
+    if 'md5sum' not in data or not data['md5sum']:
+        return
+    if 'force_md5' in request.query_string:
+        return
+    # we can of course patch / put to ourselves the same md5 we previously had
+    if context.properties.get('md5sum') == data['md5sum']:
+        return
+
+    if ELASTIC_SEARCH in request.registry:
+        search = make_search_subreq(request, '/search/?type=File&md5sum=%s' % data['md5sum'])
+        search_resp = request.invoke_subrequest(search, True)
+        if search_resp.status_int < 400:
+            # already got this md5
+            found = search_resp.json['@graph'][0]['accession']
+            request.errors.add('body', 'File: non-unique md5sum', 'md5sum %s already exists for accession %s' %
+                               (data['md5sum'], found))
+    else:  # find it in the database
+        conn = request.registry['connection']
+        res = conn.get_by_json('md5sum', data['md5sum'], 'output_file')
+        if res is not None:
+            # md5 already exists
+            found = res.properties['accession']
+            request.errors.add('body', 'File: non-unique md5sum', 'md5sum %s already exists for accession %s' %
+                               (data['md5sum'], found))
+
+
+def validate_processed_file_produced_from_field(context, request):
+    """validator to make sure that the values in the
+    produced_from field are valid file identifiers"""
+    # skip validator if not file processed
+    if context.type_info.item_type != 'output_file':
+        return
+    data = request.json
+    if 'produced_from' not in data:
+        return
+    files_ok = True
+    files2chk = data['produced_from']
+    for i, f in enumerate(files2chk):
+        try:
+            fid = get_item_or_none(request, f, 'files').get('uuid')
+        except AttributeError:
+            files_ok = False
+            request.errors.add('body', 'File: invalid produced_from id', "'%s' not found" % f)
+            # bad_files.append(f)
+        else:
+            if not fid:
+                files_ok = False
+                request.errors.add('body', 'File: invalid produced_from id', "'%s' not found" % f)
+
+    if files_ok:
+        request.validated.update({})
```

Comment:
Since we're trying to have shared code via encoded-core, consider refactoring in encoded-core to make these validators shared by passing in the item type name instead of copy-pasting with minor change here. 

Can also consider cleaning some of this up and/or removing parts that don't really make sense for this portal (like checking `produced_from` ?).

Reply:
We can't refactor to use in encoded-core at this time because the original definition hard codes the type ie: in encoded-core, it has the CGAP implementation and hard codes processed_file, hence why this didn't work originally. I can refactor this eventually though.

Reply:
Yeah, I thought it would be cleaner to refactor in encoded-core to pass the type so could be re-used easily in smaht.